File: src/omnia/data/structure/observable/ObservableSet.java
package omnia.data.structure.observable;

import io.reactivex.rxjava3.core.Observable;
import omnia.data.structure.Set;

public interface ObservableSet<E> extends ObservableDataStructure, Set<E> {

  @Override
  ObservableChannels<E> observe();

  interface ObservableChannels<E> extends ObservableDataStructure.ObservableChannels {

    @Override
    Observable<? extends Set<E>> states();

    @Override
    Observable<? extends ObservableSet.MutationEvent<E>> mutations();
  }

  interface MutationEvent<E> extends ObservableDataStructure.MutationEvent {

    @Override
    Set<E> state();

    @Override
    Set<? extends ObservableSet.SetOperation<E>> operations();
  }

  interface SetOperation<E> {

    E item();

    static <E> Observable<AddToSet<E>> justAddToSetOperations(
        Observable<? extends SetOperation<E>> observable) {
      return observable.flatMap(
          mutation -> mutation instanceof AddToSet<?>
              ? Observable.just((AddToSet<E>) mutation)
              : Observable.empty());
    }

    static <E> Observable<RemoveFromSet<E>> justRemoveFromSetOperations(
        Observable<? extends SetOperation<E>> observable) {
      return observable.flatMap(
          mutation -> mutation instanceof RemoveFromSet<?>
              ? Observable.just((RemoveFromSet<E>) mutation)
              : Observable.empty());
    }
  }

  interface AddToSet<E> extends SetOperation<E> {}

  interface RemoveFromSet<E> extends SetOperation<E> {}
}

File: src/omnia/data/structure/observable/ObservableSet.kt
package omnia.data.structure.observable

import com.badoo.reaktive.observable.Observable
import com.badoo.reaktive.observable.mapNotNull
import omnia.data.structure.Set

interface ObservableSet<E : Any> : ObservableDataStructure, Set<E> {

  override fun observe(): ObservableChannels<E>

  interface ObservableChannels<E : Any> : ObservableDataStructure.ObservableChannels {

    override fun states(): Observable<Set<E>>

    override fun mutations(): Observable<MutationEvent<E>>
  }

  interface MutationEvent<E : Any> : ObservableDataStructure.MutationEvent {

    override fun state(): Set<E>

    override fun operations(): Set<out SetOperation<E>>
  }

  interface SetOperation<E : Any> {

    fun item(): E

    companion object {

      fun <E : Any> justAddToSetOperations(operations: Observable<SetOperation<E>>):
          Observable<AddToSet<E>> {
        return operations.mapNotNull { it as AddToSet<E> }
      }

      fun <E : Any> justRemoveFromSetOperations(operations: Observable<SetOperation<E>>):
          Observable<RemoveFromSet<E>> {
        return operations.mapNotNull { it as RemoveFromSet<E> }
      }
    }
  }

  interface AddToSet<E : Any> : SetOperation<E>

  interface RemoveFromSet<E : Any> : SetOperation<E>
}